### PR TITLE
fix norm_A

### PR DESCRIPTION
**Pull request by EnLAI111**

### CODE REVIEW

File: solvers/ChambollePockPDsplit.py [hunk 1]
```diff
@@ -67,3 +67,14 @@ def _prox_huber(self, u, mu):
...
         )
+
+    def get_l2norm(self, A, n_iter=100):
+        if isinstance(A, np.ndarray):
+            return np.linalg.norm(A, ord=2)
+        else:
+            AtA = A.T @ A
+            x = np.random.randn(AtA.shape[0])
+            for _ in range(n_iter):
+                x = AtA @ x
```
Comment (agramfort) [inline solvers/ChambollePockPDsplit.py:78]: @EnLAI111 can you check if doing `(A.T @ A) @ x` is faster than `A.T @ (A @ x)`
in which case you should not create AtA

[the remark this answers is not in the document]
Comment (EnLAI111) [reply solvers/ChambollePockPDsplit.py:78]: A.T @ (A @ x) is a little faster than (A.T @ A) @ x without creating AtA

image

File: solvers/ChambollePockPDsplit.py [hunk 2]
```diff
@@ -67,3 +67,14 @@ def _prox_huber(self, u, mu):
...
+                x /= np.linalg.norm(x)
+            return np.sqrt(np.linalg.norm(AtA @ x))
```
Comment (agramfort) [inline solvers/ChambollePockPDsplit.py:80]:
```suggestion
            return np.sqrt(np.linalg.norm(A.T @ (A @ x)))
```

File: solvers/ChambollePockPDsplit.py (hunk 1 above)
Comment (agramfort) [inline solvers/ChambollePockPDsplit.py:78]:
```suggestion
                x = A.T @ (A @ x)
```

File: solvers/ChambollePockPDsplit.py [hunk 3]
```diff
@@ -67,3 +67,14 @@ def _prox_huber(self, u, mu):
             u / (mu + 1.0),
             u - self.delta * mu * np.sign(u),
         )
+
+    def get_l2norm(self, A, n_iter=100):
+        if isinstance(A, np.ndarray):
+            return np.linalg.norm(A, ord=2)
+        else:
+            AtA = A.T @ A
```
Comment (agramfort) [inline solvers/ChambollePockPDsplit.py:75]:
```suggestion
```

File: solvers/CondatVu.py [hunk 4]
```diff
@@ -69,3 +69,14 @@ def get_result(self):
...
+    def get_l2norm(self, A, n_iter=100):
+        if isinstance(A, np.ndarray):
+            return np.linalg.norm(A, ord=2)
+        else:
+            AtA = A.T @ A
+            x = np.random.randn(AtA.shape[0])
+            for _ in range(n_iter):
+                x = AtA @ x
+                x /= np.linalg.norm(x)
+            return np.sqrt(np.linalg.norm(AtA @ x))
```
Comment (agramfort) [inline solvers/CondatVu.py:82]: this function should be put in a common utils folder like done here:

https://github.com/benchopt/benchmark_resnet_classif/tree/main/utils

File: solvers/ChambollePockPDsplit.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (EnLAI111) [reply solvers/ChambollePockPDsplit.py:78]: ```A.T @ (A @ x)``` is a little faster than ```(A.T @ A) @ x``` without creating AtA

image

File: solvers/CondatVu.py (hunk 4 above)
Comment (EnLAI111) [reply solvers/CondatVu.py]: That is to say, to create a new folder ```utils``` and create a file ```l2-norm.py```?

**Review by agramfort — APPROVED** [review]

File: solvers/ChambollePockPDsplit.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomMoral) [reply solvers/ChambollePockPDsplit.py]: This really depends on the shape of `A`, so not sure this if generalizing from single observation was the right call here.

File: utils/shared.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import numpy as  np
+
+
+def get_l2norm(A, n_iter=100):
+    if isinstance(A, np.ndarray):
+        return np.linalg.norm(A, ord=2)
+    else:
+        x = np.random.randn(A.shape[1])
```
Comment (tomMoral) [inline utils/shared.py:8]: The way the operator is computed `A.T @ A` or `A @ A.T` should be decided depending on the shape of `A` no?

Comment (tomMoral) [reply utils/shared.py:8]: The factorized version is 10x faster than the non-factorized one:
image

**Pull request merged**